### DISCUSSION TITLE
Does goja have a types.d.ts?

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#3360] [Q&A] [answered]
In working with Typescript to create ph_hooks handlers, I discovered that the goja environment doesn't quite match any existing typings library. `es5.d.ts` is close of course, but then goja also supports `require()` which is found in `node.d.ts`. I've resorted to using `es5.d.ts` and appending types to `pb_data/types.d.ts` like this:

```ts
/**
 * Custom types
 */
interface Require {
  (id: string): any
}
declare var require: Require
```

```json
{
  "compilerOptions": {
    "lib": ["ES5"],
    "types": ["./types.d.ts"],
  },
}
```

Curious to hear thoughts about this. I couldn't find anything in the goja project.

**Comment by ganigeorgiev** [comment]
Goja has full ECMAScript 5.1 support + some partial ES6 features but it is not fully compliant yet (see https://github.com/dop251/goja/issues/420).

`require()` is technically not part of the ES5 spec and it is a Node.js thing and it comes from [goja_nodejs](https://github.com/dop251/goja_nodejs).

For me personally it feels overkill to go through all of the hassle of manually transpiling TS->JS and I'm not sure what to recommend.
The main idea behind the ambient TS declarations was to allow a similar code-intellisense experience (_eg. code completion and docs suggestions on hover_) with just using plain JS without the need of any complilation step, because if you are going through the trouble of running build scripts you may as well just use Go.

**Reply by benallfree** [reply]
> you may as well just use Go

For me, keeping my tech stack aligned with web development feels like the right move. And if I have TS source files that I'd like to include in the hook functionality, I need some kind of compile step anyway. Also using ES6 feature that goja might be  missing.

**Comment by benallfree** [comment, marked as answer]
Solved in #3361 

I made a starter kit for Typescript + pb_hooks https://github.com/benallfree/ts-pb-hooks-starter
